### ISSUE TITLE
Location of tooling projects [ was: Move to fsprojects? ]

### ISSUE THREAD

**Issue by dsyme** [fscheck/FsCheck#69]
Hi @kurtschelfthout 

With the FSSF becoming legally established, it probably makes sense to have http://github.com/fsharp be very strictly only those repositories related to points 1 and 2 of the [mission statement](http://fsharp.org/foundation.html) of the FSSF - e.g. fsharp/fsharp, fsharp/FSharp.Compiler.Service, the editor bindings and so on.

As a result, would it be ok to move this repository to http://github.com/fsprojects?  There are some highly respected and successful projects in that collection now, so you'd be amongst good company :)

There's a related discussion here: https://github.com/fsharp/fsharp.github.io/issues/19

Please let us know what you think.

Thanks
Don

**Comment by kurtschelfthout** [comment]
So...more than a few months have passed. Has anything new come up? To reiterate I'm happy to move Fscheck "out of the way" of the fsharp organization, as I do agree it's probably better to keep F# compiler etc there only. I'm just not too sure about the fsprojects destination.

**Comment by sergey-tihon** [comment]
It seems that we need to back to this question again. We need move votes here to make right decision

Question: Should we create new organization (`fstools` for example) for "infrastructure" projects (FAKE, Paket, F# Formatting, FsCheck, FsUnit & ProjectScaffold) or not? Please share you thoughts.

//cc: @forki @dungpa @vasily-kirichenko @ovatsus @7sharp9 @dmitry-a-morozov @isaacabraham @pblasucci @pezipink @ReedCopsey

P.S. We already have following groups:
 - [fsharp](https://github.com/fsharp) - F# Compiler and Components (Open Edition)
 - [fsprojects](https://github.com/fsprojects) - F# Community Project Incubation Space
 - [fslaborg](https://github.com/fslaborg) - Data Science and Machine Learning with F#
and I think that FAKE, Paket, F# Formatting are mature projects that able to form new org outside of fsprojects.

**Comment by tpetricek** [comment]
+1 - I'd be happy to move F# Formatting to `fstools` - I didn't move it to `fsprojects`, because I felt like `fsprojects` is a collection of fairly assorted stuff (some great, some very early stage) and so having a more high-quality organization where it would logically fit would be nice.

**Comment by kurtschelfthout** [comment]
An idea - how about a more general fsharp.contrib, basically a place where projects can "graduate" to from fsprojects? We can still have fstools and fslab for more focused organizations around a particular theme. 

Shouldn't there be a place where the more mature projects from fsprojects can go to without having to pick a theme - or that don't fit in any particular theme - or that eventually may graduate into FSharp.Core even. E.g. I would love to see more immutable collection implementations.

**Comment by forki** [comment]
Also I'm not completely sure if we need to keep the "F#" in the name.
FSharp.Formatting, FAKE, paket, projectscaffold and fscheck are tools that
work for all. NET and mono. They are just implemented in F#. If we teaser
the tools as .NET tools we will have much more users and some of the new
users will try to understand how things work internally.

**Comment by 7sharp9** [comment]
I didn't even know there was a fslaborg, having too many different orgs runs the risk of things getting lost, I think a more delayed approach to things getting into fsprojects would of been better, e.g. incubate outside fsprojects and then bring them in once there are several maintainers and stability etc.

**Comment by 7sharp9** [comment]
fsharp.charting seems to be in both fsharp and fslaborg.  

Fake is in fsharp as is FSharpMDXS, are they both tools, they both would be classed as "tooling" I guess.

Categories are hard :-)

**Comment by forki** [comment]
No please don't change semantics of existing orgs. We already did this for
fsharp and now FAKE is an alien there. This would happen to most of the
projects in fsprojects if we set the entry bar higher.

**Comment by 7sharp9** [comment]
I always thought the semantics of the fsharp org part was core engineering endorsed things that were important to be maintained.

**Comment by forki** [comment]
Yes, but this was applied after core engineering group was founded. FAKE is there for like 5 years or so. Now I think people are right to say it doesn't belong there.

**Comment by forki** [comment]
Also I wonder how do we make such decisions in the future. How can we decide if a project is mature enough for an endorsement. Personally I see two options: let @dsyme in dictatorship decide or let the fsharp foundation board decide. Both are equally good for me.

**Comment by kurtschelfthout** [comment]
Agree with a lot of what's been said (unnecessary F# focus for some projects, categories are hard,...) Feels like we're over thinking. Maybe we should just let each set of maintainers choose where they want to go/stay while mainly just cleaning up the existing groups. If a bunch of maintainers then feel like they should come together and create an organization, fine. If not, fine too.

**Comment by 7sharp9** [comment]
@kurtschelfthout I can sympathise with you wanting to move fscheck back, often organisation controlling can make you feel like a project has slipped through your fingers.  I feel like that with some of the project that I helped to create that I am no longer even an admin on.

**Comment by kurtschelfthout** [comment]
It's not really about control...I would actually love for someone to be an admin on FsCheck so I can just contribute stuff ;) it just feels like this discussion is never going to end.

**Comment by tpetricek** [comment]
This discussion definitely feels a bit unfocused. I'll let someone else do the constraint solving - here is what I'd like to do about F# Formatting:

 - I'd be happy if it shares place with ProjectScaffold, Paket and FAKE (preferably without F# focus!)
 - I'm equally happy to leave it in my repo
 - I'm not keen on having it in an org repo with lots of unrelated projects 
 - I don't want to be owner or primary maintainer of another org
 - I don't think it should be "owned" by FSSF (it should primary focus on compiler + core libs)

**Comment by ploeh** [comment]
For a while, I've been watching this from the side, but it's something I've wanted to take up.

Apart from the visibility, there's no reason for OSS to be organised within any sort of affiliation group under a language umbrella. What C# OSS is out there has 'prospered' without any shared umbrella for many years:

- NHibernate
- Castle Windsor
- Moq
- xUnit.net
- etc.

In fact, for those projects, it would have been quite laughable if they had been organised under any sort of *csharp* organisation. This is emphasised with the abundance of *choice* there is: There's not only Castle Windsor, but also Autofaq, Ninject, etc. There's not only Moq, but also NSubstitute, FakeItEasy, Foq (written in F#!), etc. There's not only xUnit.net, but also NUnit, Fixie, etc.

As [Oren Eini quite astutely pointed out recently](http://ayende.com/blog/170849/why-ravendb-isnt-written-in-f-or-the-cost-of-the-esoteric-choice), it makes F# look immature that everything is gathered in the same organisation. Personally, I think it would strengthen the language with a diaspora of OSS. This would also open the floor for competition.

Not that I'm particularly unhappy with the current selection of F# tooling, but as an example, there seems to me to be two up-and-coming web frameworks (or are they *libraries?*): Suave and Freya. Should they both go into the same organisation? Is there room for both of them? Or wouldn't it be better if they were allowed to evolve independently (as they currently do)?

**Comment by 7sharp9** [comment]
To be fair I've never particularly liked grouping things here, I'm quite happy for things to be just in GitHub somewhere.  I think it makes sense for core important things like how Clojure incubate small intrinsic projects but there's no formal management of that here just a loose collection of "stuff".  Maybe that's what the fsharp repo is anyway though. 

D.

**Comment by kurtschelfthout** [comment]
As a follow up, as #100 indicates, I've gone ahead and created an fscheck organization. I intend to transfer the fscheck repository there in the next few days.

I'm hopeful this is the right choice for fscheck (definitely more so than moving to my own profile - try remembering my last name as part of a url... :) ) as well as sorting things out in the fsharp organization.

Please don't take this the wrong way - I'm not going anywhere - and none of this is due to what anyone in the F# community has done or said, and I am certainly grateful for the support. You folks have been awesome, and I hope you continue to be!

**Comment by tpetricek** [comment]
Sounds like a good decision to me! I think FsCheck is a project that's mature enough to benefit from standalone organization.

With a separate organization for FsCheck, it sounds like you should actually be able to host the docs at http://fscheck.github.io, which would be cool :-)

**Comment by dsyme** [comment]
This is a good move, both for FsCheck and github.com/fsharp, great work @kurtschelfthout!! 

@ploeh - I personally analyse the C# open source situation very differently, and in particular I think that the comment in the blog post you mentioned missed the mark by a fair way. Indeed, so much so that I think C# should be learning from F# here, rather than the other way around :) This probably isn't the place to have this discussion in full, but I'll put some thoughts here in any case. 

First, C# and openness.... As much as some C# OSS projects prospered during the last 10 years (and well done to them!), during that time C# and .NET unfortunately accumulated an unfair reputation (especially among outsiders) of being a predominantly closed-source ecosystem. I think it's important for the thought leaders in the C# and .NET world to ask the question "How did this happen? How did we end up with this reputation? What can we do about it?".  

To some extent, the answer lies with Microsoft's old approaches, now gone.  However, that's only part of the story.  In my opinion, this could reputation have easily been very different if only there had been some kind of FSSF-like grassroots initiative centered on openness which owned the message about C# and .NET openness and helped facilitate it.  That is, the C# world (particularly MVPs and other key contributors) should have put forward a strong message that C# was welcoming to open source.  There was some great work in this regard - and all power and respect to those who did it - but until recently there has been no widespread grassroots C# or .NET community organization presenting a voice for C# or .NET and openness.

If such an initiative had been in place, I think so many good things would have happened and the whole reputation of C# and .NET as an open ecosystem and community would have been better. Now that C# is "open and cross-platform from the ground up"  we're finally seeing things shift in this direction.

If you start with this POV, and imagine what the C# open world _could_ have looked like, including its reputation and self-empowered community dynamic, then it's pretty reasonable to imagine that parts of the C# open community would have done much as the F# community have recently done: those leading this effort would have found innovative ways to share news about open projects, incubate projects together and welcome new projects to the fold. Who knows what umbrella organizations would have formed to support those initiatives.  But one thing I'm sure of: people would have pointed to those umbrellas as social proof of the fact that C# embraced openness.  A couple of isolated projects don't convince anyone - an active social organization does.

There are hundreds of ways that might have played out in practice, but my point is that if C# had actually had a very widespread open ethos and spirit in the last 15 years, then the whole C# open project world would have looked different (and possibly a bit more like the F# or Python or Java worlds). The world of individual, isolated successful open projects (only brought together late in the day under nuget.org) is a symptom of an underlying problem, rather than a positive to aspire to.    Indeed we're now starting to see change with the collection of many C# open projects under the .NET Foundation. That's really an indicator that C# and .NET are moving in the direction of umbrella organizations with more of a grassroots feel.  

The F# world is active and welcoming to openness, and efficient in its sharing and communication of open projects.  The "mature" C# world, in contrast, unfortunately developed an undeserved reputation for being based at its core on closed source, proprietary software, and newcomers did not historically find it welcoming to openness (again a largely undeserved reputation to be sure).  

C# is where it is and these things tend to be a bit historical and don't change quickly. It's possible that the future of organized C# openness may start to look more like the F#, Python or Java worlds. Are C# and .NET immature because they only just got the .NET Foundation? Yes, in some ways - a deep embrace of openness is new to much of the C# and .NET community. The C# world is now having the humility to recognize that.

In any case, personally I think the F# world has got a good grip on open projects.

* First, so much has happened in the last 4 years.  It's great.

* Second, the core github.com/fsharp projects are now a pretty clean, minimal, orthogonal set of projects (a couple of small improvements still need to be made). 

* Third, the F# world is very welcoming to new open projects and socializes them on twitter and elsewhere (regardless of what umbrella they live under).

* Fourth, the F# world seems to have found many positive ways to incubate new projects, either alone or cooperatively, be it http://github.com/fslaborg, http://github.com/fsprojects, new organizations like http://github.com/suaveio or http://github.com/mbraceproject, WebSharper projects or the like.  I think these umbrella organizations serve a very positive role and are being largely successful, and there is room for creating more of them. For example, I think it would be very positive to have a places to cooperatively incubate WebSharper projects, or FsLab projects, or MBrace projects  - newcomers and old-timers alike learn so much about how to engineer and deliver projects when they see what other people are doing.  

* Fifth, the community are good at making the structural adjustments as things grow - like @kurtschelfthout is doing here (or, similarly, when I [suggest  people rename projects and namespaces to make things fit more nicely](https://github.com/fsprojects/FSharp.Text.RegexProvider/issues/11)) they nearly always respond positively and quickly).

> Not that I'm particularly unhappy with the current selection of F# tooling, but as an example, 
> there seems to me to be two up-and-coming web frameworks (or are they libraries?): 
> Suave and Freya. Should they both go into the same organisation? Is there room for both 
> of them? Or wouldn't it be better if they were allowed to evolve independently (as they currently do)?

The rule I see being followed is "let many flowers bloom, find ways to support their growth".  For frameworks (WebSharper, Suave, MBrace, ...) and some tools (FsCheck) a new organization is often used.  For core libraries (and some tools) people seem to find "fsprojects" or "fslaborg" useful, and some other libraries live under a company umbrella (Deedle).

Anyway, in summary, I think the F# world of open projects is in relatively good shape and it is supporting the needs of F# users pretty well. There's lots of improvements to make, but reverting to a world where no umbrella organizations exist besides individuals and companies would not be step forward for F#, nor is it where C# needs to be going forward.

**Comment by haf** [comment]
@dsyme Well written!

@ploeh I just wanted to chime in re. the comments on Suave/Freya:

We are having a discussion here https://github.com/SuaveIO/suave/issues/153 if you're interested it would be great to have you partake in it.

For me it's about compositionality of web frameworks; I want to move Suave towards more of that. If you just want it as a HTTP server, go for it and run Freya on top and use its Machine. Or use Suave mainly for WebSockets/SSE and distributed tracing/APIs and Freya on the side. Or help out to finish the OWIN branch for Suave, and you can e.g. run @leastprivilege's IdentityServer on top. =)

A thousand flowers lead to innovation, but take time to pick. Especially for newbie flower-pickers. And time to properly secure, because effort is spread out. Having ongoing and well-considered changes/improvements done to existing projects, arising from the buzz of cross-pollination (pun intended), is probably the way to go here.

**Comment by panesofglass** [comment]
@dsyme very nicely done!

@ploeh, I concur with @haf on the web libraries. Our goal with Freya is to create a stack of libraries, building on OWIN, that allow you to do what you want. You could even run WebSharper on top of Freya.Core using [WebSharper.Owin](https://github.com/intellifactory/websharper.owin) on top of Suave's web server (once OWIN support is complete). The endgame is to enable developers to use what they want, where they want. I'm also currently working with the ASP.NET team to see if we can share some of their tutorial content to make it easier to offer well-written, good examples of building web apps.
